[RDY] Fix xfree of static value from vim_getenv() refactor

mbainter opened the pull request:
Fixes #2487.

The refactoring of vim_getenv() to remove the mustfree arg
included reworking calling functions.  expand_shellcmd was
also using that to track its usage of the variable within
the function, resulting in #2487. This change addresses
that scenario and cleans up some of the function for style.

mbainter committed:
Fix xfree of static value in expand_shellcmd()

The refactoring of vim_getenv() to remove the mustfree arg
included reworking calling functions.  expand_shellcmd was
also using that to track its usage of the variable within
the function, resulting in #2487. This change addresses
that scenario and cleans up some of the function for style.

elmart commented:
Thanks for fixing this so quick.
Couple of things:
- Please take as a habit separating style-only changes and meaningful changes. In this case, you can make a "Cleanup" commit with all style changes, and a "Fix" commit with the real fix.
- It would be nice if you reviewed your last merged PR, so that we make sure there are no other cases like this. I mean, review calling places where `mustfree` was removed, to make sure we don't have more static strings incorrectly freed. I think to remember there weren't many places, and it's strange that error passed reviewers filter (my fault), but just to be sure. If there's some more case, please include it in this PR too.

elmart commented:
BTW, we use [WIP] (work in progress) for what you have used [InProgress]. Convention is [WIP] --> [RFC] --> [RDY]. Bot automatically assigns tags if you follow it, and it's clearer for everyone else.

mbainter commented:
No problem.  I marked it as in progress as I want to go through the rest just to be sure I didn't miss that anywhere else.  When I made the change I was looking for that, but somehow I missed it here.  I want to be sure I didn't anywhere else before you merge this.  I wanted to get the fix out here right away though.

Good point on the style vs fix, I should know better than that.  I will make sure I separate them going forward.

elmart commented on src/nvim/ex_getln.c:
```diff
@@ -3832,49 +3832,56 @@ static void expand_shellcmd(char_u *filepat, int *num_file, char_u ***file,
-  else if ((pat[0] == '.' && (vim_ispathsep(pat[1])
-                              || (pat[1] == '.' && vim_ispathsep(pat[2])))))
+  } else if ((pat[0] == '.' && (vim_ispathsep(pat[1]) 
+        || (pat[1] == '.' && vim_ispathsep(pat[2]))))) {
```
This doesn't follow styleguide. `||` should be columned below `vim_ispathsep`, as before.

elmart commented on src/nvim/ex_getln.c:
```diff
@@ -3882,24 +3889,28 @@ static void expand_shellcmd(char_u *filepat, int *num_file, char_u ***file,
   xfree(buf);
   xfree(pat);
-  xfree(path);
+  if(mustfree) {
```
Space needed after `if`.

mbainter commented:
Okay, I went through everywhere vim_getenv() is called in the codebase, and reviewed the use of the variable it returns.  There was only one other place that variable is reused in that way, and it already had a mustfree variable wrapped around  that use.  I think this PR is ready.

(I addressed the issues raised above as well)

elmart commented:
LGTM.

elmart commented:
Ok, some more new contributor training:   ;-)
- I added "Fixes #2787." to your first comment.  That way, when this gets merged, issue is automatically closed. Please do so in future when fixing an open issue.
- You should not mark RDY before all your reviewers have given their LGTM. RDY is there for PR's that can just be merged without further thinking if maintainer sees one.

Waiting for build to pass and then I'll merge. Thanks!

oni-link commented on src/nvim/ex_getln.c:
```diff
@@ -3841,6 +3841,7 @@ static void expand_shellcmd(char_u *filepat, int *num_file, char_u ***file,
 
   flags |= EW_FILE | EW_EXEC;
 
+  bool mustfree = false;  // Track memory allocation for *path.
```
Why use`*` in the comment?

elmart replied:
Ooops, sorry, I just saw your comment right after merging.
No big deal, though.
